### PR TITLE
Automatically build Docker images for branch and tag pushes

### LINKED ISSUE #901: Update CI/CD to handle main builds, and document main build availability
CI/CD should handle producing builds from main. #900 was wrong, but there's not yet any right target to point it to. We need a main (and possibly next?) build, but currently only have some outdated master builds.

We reference these in docs, or rather previously referenced builds off master, and those appear to have gotten lost in the master->main transition.

Follow-up from #399 

Desired end state is that we produce builds from main, and docs point to those builds.

### PR DESCRIPTION
**Pull request by rainest**
**What this PR does / why we need it**:
Adds GitHub Actions workflows for automatically building images on main/next branch and tag pushes.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

Provides a controller Red Hat artifact for https://github.com/Kong/team-k8s/issues/33

Fixes #901 

**Special notes for your reviewer**:
I haven't actually created this repo in the Docker Hub Kong org yet. Figured I'd confirm this is what we want to do first.

This requires Docker credentials in a GitHub repo secret. I have access to add a token, but not access to create repo secrets, so we'd need to have a repo admin do that. Do not merge until this and repo creation are complete.

Not sure if there's a better way (i.e. something we can add in the Dockerfile itself) to handle making Alpine the default target that doesn't rely on the `--target` argument in the make target. However, since using `make container` is our standard way to manually build images, that approach seems fine.

This doesn't yet handle RC images, and we'd probably want it to. I'm trying to think of a way to do that without creating a third workflow, since the only difference would be that we don't push latest.

For practical examples, see the [branch builds](https://github.com/rainest/kubernetes-ingress-controller/actions/runs/712395020) and [tag builds](https://github.com/rainest/kubernetes-ingress-controller/actions/runs/712443588) (ignore the workflow name, it was incorrect when I tested) on my fork.

Outputs can be seen on [my Docker Hub repo](https://hub.docker.com/repository/registry-1.docker.io/traines/kic/tags?page=1&ordering=last_updated). The names are a bit different from the final version.

### CODE REVIEW

File: README.md
Comment (shaneutt) [inline README.md:132]: That link requires login does it not? should we point to the org instead? https://hub.docker.com/u/kong

**Review by shaneutt — APPROVED** [review]
This seems like a good starting point, and we can always tune things further after the fact.

File: README.md
Comment (rainest) [reply README.md]: Indeed. Didn't realize that was a private URL, and hadn't created the repo yet, so double-whammy bad link. Having no created the repo, 19ee1fd

**Comment by rainest** [comment]
Thinking over how to handle RC images some more, I realized that tagging latest automatically doesn't work for the non-RC images either, since we may backport fixes and release patch releases for older versions also. As such, changing the tag job to just handle all git tags, but only push the images for those tags specifically in 0faa158.

We'll need to tag Docker `latest` manually when appropriate, since there's nothing in the tag itself that'd allow us to use workflow filters to distinguish when a tag is the newest available version versus a backport.

**Review by shaneutt — DISMISSED** [review]

**Review by mflendrich — COMMENTED** [review]
Attached some technicalities and thoughts about how orthodox our CI process should (or should not) be.

File: .github/workflows/docker-release.yaml
Comment (mflendrich) [inline .github/workflows/docker-release.yaml:35]: Like said in the `Makefile` comment, I recommend that we use the makefile (instead of this action) to build the image, to ensure that the exact same build method is used by CI and manual workflows. This action can be used for the push later.

File: Dockerfile
Comment (rainest) [reply Dockerfile:23]: It's a Red Hat requirement for listing on their registries. This PR doesn't push to them, though I suppose I should add that for the release builds given that we'll have the image for it.

Their docs aren't easily navigable and some parts are hidden behind a login wall. That

> Test name: has_licenses
Why? So the end user is aware of the terms and conditions applicable to the software. Including opens source licensing information, if open source components are included in the image.
How? Create a directory named /licenses and include all relevant licensing and/or terms and conditions as text file(s) in that directory.

The closest public equivalent I can find is https://redhat-connect.gitbook.io/partner-guide-for-red-hat-openshift-and-container/program-on-boarding/containers-with-red-hat-universal-base-image-ubi and [the linked examples](https://github.com/RHC4TP/starter/tree/master/Container%20Zone/UBI-Example).

Comment (mflendrich) [reply Dockerfile:23]: I see, thanks for explaining the context to me.

I think it's a more complex problem than it seems: the `LICENSE` file is most probably not enough - licenses such as Apache 2.0 require redistribution of copyright notices (usually in the form of a `NOTICE` file) as well. It may be so that there are no such notices today but they will be added in the future, posing a risk of accidental omission.

To keep the scope manageable, I recommend that we treat satisfying RH's license needs as a separate problem (by creating a separate issue to track sufficient introduction of license data to docker images - @eskibars may probably help here). Any (potentially) half-baked solution here will suppress the check on RH's side but potentially omit something important.

File: .github/workflows/docker-release.yaml
Comment (rainest) [reply .github/workflows/docker-release.yaml]: Yeah, leftover from a previous iteration that didn't modify the original Dockerfile. dbc194f

File: .github/workflows/docker-dev.yaml
Comment (rainest) [reply .github/workflows/docker-dev.yaml:1]: The repo these jobs push to is for public consumption, and discussion with Harry raised what I think is a legitimate point: having a proliferation of tags there is confusing. Users shouldn't have any reason to pull individual commit releases. They should almost always be running release builds (builds for Git tags) and may want to test pre-release builds (e.g. we solve a bug, merge the fix, and ask them to try the latest main/next tag to confirm the fix without having to build and push their own image somewhere).

Test CI jobs have a different set of needs, as they're interested in individual commits on any branch, and almost all of the artifacts (anything that doesn't end up being on the tip of main or next) aren't of interest outside their specific test run. As such, I like the approach taken in the Railgun test (just build and run the Go binary) since the jobs clean up after themselves. Generating and uploading image artifacts would be more appropriate if we needed to trigger tests outside the controller repo, but as of yet we don't.

I recognize the interest in having the same artifact regardless, but divergence there hasn't been a source of issues for us. There's not really anything of significance in the images beyond the Go binary: they're basically just necessary packaging to provide something K8S can run that includes version metadata.

The current split between the jobs is a simple means of handling the different sources for the Docker tag names, so that they're identical aside from the `on` criteria and output in the `extract_shortref` step and still simple and straightforward. Unifying them is likely possible, but requires a lot more complex conditional logic and manipulation of step outputs to produce the final tag list. To Shane's point earlier, this is a start, and we can expand on it later.

File: .github/workflows/docker-release.yaml
Comment (rainest) [reply .github/workflows/docker-release.yaml:35]: Moving to Docker's action was partially to benefit from upstream work and partially because its build caching just works--the Go binary build time is far longer than the other stages. Not sure if that's easily doable with the Makefile alone.

The Makefile target doesn't do anything other than populate basic values for the `docker build` invocation, and we can gather the same info/store it in the environment equally well with the action, so I don't see a concern for the difference in the manual workflow.

Comment (hutchic) [reply .github/workflows/docker-release.yaml:35]: use buildkit and you can benefit from `RUN --mount=type=cache,target=/root/.m2 mvn -o install` for cacheing purposes

[prior art](https://github.com/Kong/docker-pulp/blob/main/pulp-core/Dockerfile#L40)

I agree with the sentiment of having the local way of doing things being equivallent to the CI way of doing thing. An option would be to have the logic in github actions and tell Make to use it

```
.github/workflows/docker_build.yml
.github/workflows/docker_release.yml

docker_build:
    act -W .github/workflows/docker_build.yml -v
```

File: .github/workflows/docker-dev.yaml
Comment (hutchic) [reply .github/workflows/docker-dev.yaml:1]: I've had good experiences with https://github.com/marketplace/actions/docker-meta

TLDR 
image

File: .github/workflows/docker-dev.yaml
Comment (mflendrich) [inline .github/workflows/docker-dev.yaml:64]: One last question.

We seem to be in agreement that development images should not bloat the user-facing image repository with tags and that these should go to a separate repo.

At the same time, `main` and `next` are clearly development tags - it just happens that more will follow as we improve our CI.

Maybe the right moment to establish that separate repo for dev tags is now (so that `main` and `next` go there)? If, otherwise, we do this later, we'll need to remove `main` and `next` tags from this (user-facing) repo.

**Review by mflendrich — COMMENTED** [review]
As discussed over zoom, we'll start small and handle requests from review comments as separate issues.

One last doubt regarding image repositories involved.

File: .github/workflows/docker-dev.yaml
Comment (rainest) [reply .github/workflows/docker-dev.yaml:64]: See https://github.com/Kong/kubernetes-ingress-controller/pull/1156#discussion_r610919251

They are a limited set of pre-release tags for branches that contain code that will eventually be released. They're there if users wish to test new features and/or stage configuration in their dev environments that requires unreleased functionality so that they can more quickly promote to prod once a release tag is cut. There are only two of them and it's easy to understand what they represent, whereas the individual ref builds used in CI are many, require you know which hash corresponds to a particular feature, and may not be fully reviewed.

Comment (mflendrich) [reply .github/workflows/docker-dev.yaml:64]: The way I see the state we're heading towards is:

1. some docker registry A has release tags `1.0`, `1.1.1`, `2.0`, `2.0.0-alpha.1`, `2.0.0-beta.3` etc that exist there indefinitely
2. some docker registry B has tags created from git SHAs, from every build - that are immutable, but are pruned after some time (e.g. a month)
3. some docker registry C has tags created from branch names `main`, `next`, `feat/awesome-feature`, etc etc etc which are a (moving) alias for the git SHAs above

I was thinking that `kong/kubernetes-ingress-controiller` makes sense for A, and something like `kong/kubernetes-ingress-controller-dev` for B and C.

Would you like to (in the long run) replicate `main` and `next` into A as well? Or split C into `main`/`next` going into one repo, and all branches BUT these two going to another repo?

Comment (rainest) [reply .github/workflows/docker-dev.yaml:64]: > Would you like to (in the long run) replicate main and next into the "user-facing" repo as well?

This is what I intended to do from the outset.

Ultimately it's not difficult to add an additional push target--you just add another item under the `tags` section that uses another repo, so I wouldn't worry about this for now.

**Comment by rainest** [comment]
Credential secrets are now in place, so this is clear to go in once we're happy with it. Removed do not merge label.

**Comment by rainest** [comment]
Based on Colin's suggestion, added a unified workflow in 3f9fdb3. This is intended to ultimately replace the others, but I haven't removed those yet to preserve existing comments (we should remove them before merging).

https://github.com/rainest/kubernetes-ingress-controller/runs/2423767024 is an example tag push run.
https://github.com/rainest/kubernetes-ingress-controller/runs/2423764803 is an example branch push run.

During the course of setting that up, I realized that the TAG argument handling in the workflow differed from the current behavior, sort of. In the Makefile targets, `TAG` is a static value (normally set to the newest tag) at the top of the Makefile. Builds of unreleased content (e.g. the tip of main) still indicate that their version is 1.2.0 or whatever. Earlier versions of the workflows used the same version as used in the Docker tag (e.g. the image would have the label `version=main` for those builds). Do we have a preference for one or the other? I added a step to fetch the most recent tag in the interest of preserving existing behavior, but personally feel the existing behavior is less appropriate for the main and next builds.

**Comment by mflendrich** [comment]
> Do we have a preference for one or the other? I added a step to fetch the most recent tag in the interest of preserving existing behavior, but personally feel the existing behavior is less appropriate for the main and next builds.

My preference is that if an image gets built from a source that's different than the source for a well-defined tag (say. `1.2.0`) then its tag should not be 1.2.0 to avoid confusion. So I opt for dropping the existing behavior.

**Review by mflendrich — COMMENTED** [review]
> Do we have a preference for one or the other? I added a step to fetch the most recent tag in the interest of preserving existing behavior, but personally feel the existing behavior is less appropriate for the main and next builds.

My preference is that if an image gets built from a source that's different than the source for a well-defined tag (say. `1.2.0`) then its tag should not be 1.2.0 to avoid confusion. So I opt for dropping the existing behavior.

Let me know when the PR is ready for review with all the stuff to be removed actually removed :+1:

**Comment by rainest** [comment]
@mflendrich raise a concern with the major/minor Docker tags applying to rc/alpha/etc. releases during Wednesday's sync, but I believe these are fine. Consider the following order of releases:

1. We push a `1.2.1` git tag. CI tags this image `kic:1.2.1` and `kic:1.2`.
2. We push a `2.0.0-alpha.1` tag. CI tags this image `kic:2.0.0-alpha.1` and `kic:2.0`
3. We push a `1.2.2` git tag. CI tags this image `kic:1.2.2` and `kic:1.2`, overriding the previous `1.2` tag, which was an alias `1.2.1`.
4. We push a `2.0.0` tag. CI tags this image `kic:2.0.0` and `kic:2.0`, overriding the previous `2.0` tag, which was an alias `2.0.0-alpha.1`.

This does not inadvertently upgrade you to a pre-release version if you are using the major-minor tags: if you've set the current stable major-minor tag (`1.2`), you'll only get 1.2.x patch releases. To get `2.0.0-alpha.1`, you'd need to intentionally upgrade to `2.0`.

Major-minor tags would downgrade you if we performed a release out of order, e.g. if there was a step 5 that released `2.0.0-alpha.2`. However, I see no reason that we'd do this--once we've pushed out a stable `2.0.0`, we have no reason to release additional pre-release versions after. All subsequent 2.0 releases should be stable patch releases.

Automatic `latest` tagging _would_ present an issue. As far as I know, docker-meta has no logic beyond just tagging `latest` when it's enabled, so `latest` always points to the artifact generated by the CI run that completed last. Using it (although bad practice) would upgrade you to `2.0.0-alpha.1` from `1.2.1` at step 2, and then downgrade you to `1.2.2` at step 3, which we don't want.

Automatic major tags would present a similar issue if we released a `1.2.1` and then a `1.3.0-alpha.1` or similar, but we don't use automatic major tags because even a stable release on them (e.g. `1.2.1` followed by `1.3.0`) could contain breaking changes; so we wouldn't want users to receive those updates automatically.

**Review by shaneutt — APPROVED** [review]
:heavy_check_mark: once @mflendrich's comments are resolved and he's :heavy_check_mark:

**Comment by rainest** [comment]
@mflendrich ping following https://github.com/Kong/kubernetes-ingress-controller/pull/1156/commits/b0d43dd2741e5e9cd9c40901f6da33535fe3fd6e

**Review by mflendrich — APPROVED** [review]

**Pull request merged**
